processing/src/main/java/io/druid/segment/incremental/OakIncrementalIndex.java

```diff
@@ -0,0 +1,569 @@
+
+    if (numEntries.get() < maxRowCount || skipMaxRowsInMemoryCheck) {
+      oak.putIfAbsentComputeIfPresent(incrementalIndexRow, row, computer);
+
+      int currSize = oak.entries();
+      int prev = numEntries.get();
+      while (currSize > prev) {
+        if (numEntries.compareAndSet(prev, currSize)) {
+          break;
+        }
+        prev = numEntries.get();
+      }
+
+    } else {
+      if (!oak.computeIfPresent(incrementalIndexRow, computer)) { // the key wasn't in oak
+        throw new IndexSizeExceededException("Maximum number of rows [%d] reached", maxRowCount);
+      }
+    }
+    return numEntries.get();
+  }
+
+  @Override
+  public BufferAggregator[] getAggs()
+  {
+    return aggsManager.getAggs();
+  }
+
+  @Override
+  public AggregatorFactory[] getMetricAggs()
+  {
+    return aggsManager.getMetricAggs();
+  }
+
+  public static void aggregate(
+          AggregatorFactory[] metrics,
+          boolean reportParseExceptions,
+          InputRow row,
+          ThreadLocal<InputRow> rowContainer,
+          ByteBuffer aggBuffer,
+          int[] aggOffsetInBuffer,
+          BufferAggregator[] aggs
+  )
+  {
+    rowContainer.set(row);
+
+    for (int i = 0; i < metrics.length; i++) {
+      final BufferAggregator agg = aggs[i];
+
+      synchronized (agg) {
+        try {
+          agg.aggregate(aggBuffer, aggBuffer.position() + aggOffsetInBuffer[i]);
+        }
+        catch (ParseException e) {
+          // "aggregate" can throw ParseExceptions if a selector expects something but gets something else.
+          if (reportParseExceptions) {
+            throw new ParseException(e, "Encountered parse error for aggregator[%s]", metrics[i].getName());
+          } else {
+            log.debug(e, "Encountered parse error, skipping aggregator[%s].", metrics[i].getName());
+          }
+        }
+      }
+    }
+    rowContainer.set(null);
+  }
+
+  @Override
+  public IncrementalIndexAddResult add(InputRow row, boolean skipMaxRowsInMemoryCheck) throws IndexSizeExceededException
+  {
+    IncrementalIndexRowResult incrementalIndexRowResult = toIncrementalIndexRow(row);
+    final int rv = addToOak(
+            row,
+            numEntries,
+            incrementalIndexRowResult.getIncrementalIndexRow(),
+            in,
+            skipMaxRowsInMemoryCheck
+    );
+    updateMaxIngestedTime(row.getTimestamp());
+    return new IncrementalIndexAddResult(rv, 0, null);
+  }
+
+  private IncrementalIndexRow getMinIncrementalIndexRow()
+  {
+    return new IncrementalIndexRow(this.minTimestamp, null, dimensionDescsList, IncrementalIndexRow.EMPTY_ROW_INDEX);
+  }
+
+  @Nullable
+  @Override
+  public String getMetricType(String metric)
+  {
+    return aggsManager.getMetricType(metric);
+  }
+
+  @Override
+  public ColumnValueSelector<?> makeMetricColumnValueSelector(String metric, IncrementalIndexRowHolder currEntry)
+  {
+    return aggsManager.makeMetricColumnValueSelector(metric, currEntry);
+  }
+
+  @Override
```

Comment:
In general, it looks like there are many non-Oak-specific functions here. Consider using an abstract base class maybe?